File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # bucket sort
        
        seen = [[] for i in range(len(nums)+1)]
        counts = defaultdict(int)
        
        for i in range(len(nums)):
            counts[nums[i]] += 1
        
        for num, count in counts.items():
            seen[count].append(num)
            
        ans = []
        
        
        for i in range(len(seen)-1, 0, -1):
            for num in seen[i]:
                ans.append(num)
                
                if len(ans) == k:
                    return ans
                        
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const counts = {};
    const bucket = Array.from({length: nums.length + 1}, () => []);
    const ans = [];
    
    for (const num of nums) {
        counts[num] = (counts[num] || 0) + 1;
    }
    
    for (const num in counts) {
        bucket[Number(counts[num])].push(+num);
    }

    for (let i = bucket.length - 1; i >= 0; i--) {
        while (ans.length < k && bucket[i].length) {
            ans.push(bucket[i].pop());
        }
        
        if (ans.length === k) break;
    }
    
    return ans;
};